### PR TITLE
Change the way the shape is done in the background container

### PR DESCRIPTION
**Pull request by psychon**
Previously, the background container "just" used the shape and drew a
line around it. This means that half the line will be inside of the
shape and half of it will be outside. Thus, this hides the actual shape
that is used.

The following example images are generated for this example:
```lua
    s.mywibox.height = 40
    s.mywibox:setup {
        layout = wibox.layout.fixed.horizontal,
        {
            layout = wibox.layout.stack,
            wibox.widget.textbox("This text\nis behind"),
            {
                layout = wibox.container.background,
                shape = gears.shape.rounded_rect,
                shape = gears.shape.infobubble,
                bg = "#0008",
                fg = "#fff8",
                shape_border_color = "#00f8",
                shape_border_width = 2,
                wibox.widget.textbox("Test with some long text\nand a line break\nactually two")
            }
        },
    }
```

This commit changes that so that the line is added outside of the shape.
It does this via some tricks:

- In :before_draw_children(), :push_group() is used to redirect drawing
  of the child widget to a temporary surface. (This surface is initially completely transparent).
- In :after_draw_children() we thus get the content in a temporary surface.
out
- In :after_draw_children(), the border is added to this group.
  + For this, another temporary surface is created. It will be used as a
    mask.
  + The inside of the shape on this mask is cleared, everything else is
    filled. Thus, the mask now contains everything "not content".
out
  + Everything inside the mask is filled with the background color.
out
- Also in :after_draw_children(), the group is drawn to the actual
  target surface.
  + Again, this needs a mask.
  + This time, we draw the shape to the mask with twice the border width.
    Thus, half of this line will be outside of the shape.
out
  + Then, the shape itself is also filled so that the mask contains the
    shape and the border.
out
  + This mask is then used to copy the right parts of the temporary
    surface were the child widget and border was drawn to the actual
    target surface that will be visible on screen.
Before this, the target contains:
out
Afterwards:
out

This approach has some upsides. Because we no longer have "half the
border" above content, colors with some transparency work fine for the
border. Also, this should avoid issues with anti-aliasing, because e.g.
the border is not just drawn with the border width, but also further out
to everything else so that the background cannot "bleed through".

Fixes: https://github.com/awesomeWM/awesome/issues/2516
Signed-off-by: Uli Schlachter <psychon@znc.in>

### CODE REVIEW

**Comment by psychon** [comment]
```
lua: ../../../../src/cairo-pattern.c:1062: cairo_pattern_reference:
   Assertion `(_cairo_atomic_int_get (&(&pattern->ref_count)->ref_count) > 0)' failed.
```
Yay, another LGI bug (that somehow I did not hit locally).

Also: This commit effectively removes `shape_clip` support by always enforcing `shape_clip=true` (but in a way that does not produce the antialiasing issues that @actionless was seeing -- I finally understand were they come from!). Does anyone really want `shape_clip=false`?

**Comment by actionless** [comment]
i think shape_clip=false could be imitated easily with stack widget with actual background and with transparent background+shape

**Comment by actionless** [comment]
@psychon which lua and lgi versions you have locally?

**Comment by actionless** [comment]
mb some of the shims doing something weird?

**Comment by psychon** [comment]
It's one of the lines that I did not modify: https://github.com/awesomeWM/awesome/pull/2609/commits/028fda93438ce22e4ae5377cd4182369c985b435#diff-885682a6d2d38210025f1f3b0d402f49L45
This calls `set_source` to draw the background, but the background pattern is invalid (I guess: already freed).

So... this does not look like a new bug.

**Comment by psychon** [comment]
I pushed a work-around. Let's see what Travis says.

**Comment by Elv13** [comment]
> [15:12] <psychon> Elv13: should the "shape outside" PR be merged?

You decide. After all these problems, to me it looks scary to merge this as the last commit. Still, it is a very real and (to some) very annoying bug.

**Comment by psychon** [comment]
I just remembered that there is still the `shape_clip` problem. How should we handle that?

* Deprecate `shape_clip` because no one wants `shape_clip=false' anyway and it is just an historical accident that this option exists.
* Implement `shape_clip=false` again.

**Comment by actionless** [comment]
i would choose deprecating

**Comment by Elv13** [comment]
Add a comment in the "behavior changes" section, print a gdebug warning when it is explicitly set to `false`. Make that warning mention to move the code to a `stacked` layout. Be done with it?

**Comment by psychon** [comment]
Done (except for the "behaviour changes" section, there is no such section yet, is there?)

**Comment by psychon** [comment]
Sounds like a good idea (except that I bet we will forget to update the `NEWS`, but that doesn't mean that this is worse than the old approach). Done in #2624.

File: lib/wibox/container/background.lua [hunk 1]
```diff
@@ -268,20 +268,16 @@ function background:get_shape_border_color()
...
--- @property shape_clip
--- @tparam boolean value If the shape clip is enable
-
 function background:set_shape_clip(value)
-    if self._private.shape_clip == value then return end
-
-    self._private.shape_clip = value
-    self:emit_signal("widget::redraw_needed")
+    if value then return end
+    require("gears.debug").print_warning("shape_clip property of background container was removed."
```
Comment (Veratil) [inline lib/wibox/container/background.lua]: This is removing, not deprecating as the commit says, `shape_clip`?

Comment (psychon) [reply lib/wibox/container/background.lua]: Two commits before that (the commit that does "the big rewrite") does the actual removing. This commit just turns "silently ignore this option" into "un-silently ignore this option".
So, this is only removing the internal implementation of the setter, but its "outside observable" behaviour is basically the same (the only observable difference is that after setting the option to `false`, the getter will still return `true`; no visual differences).

Comment (Veratil) [reply lib/wibox/container/background.lua]: Ah! Understood. Just used to seeing a deprecation warning rather than a removed one. :)

**Comment by actionless** [comment]
so, since release is out, would we merge it for testing by a wider audience?

**Comment by actionless** [comment]
seems new travis is not aware of this PR

**Comment by Elv13** [comment]
> seems new travis is not aware of this PR

Yes, it is documented. It's why I waited for the release to risk pressing the migrate button. The old builds are *not* imported. Travis-CI.org and .com apparently have different databases and the import tool is documented not to import this old builds.

**Comment by actionless** [comment]
it broken background's `set_fg()` under some tricky usecase, i'll create a separate ticket once i'll extract it to be reproducible on default rc.lua

**Pull request merged**